payjoin-directory/src/db.rs

```diff
@@ -1,191 +1,105 @@
+        tokio::time::timeout(self.timeout, self.peek(pubkey_id, channel_type)).await.ok()
     }
 
-    // Step 2: If no content was found, set up a listener
-    let mut listener = PgListener::connect_with(pool).await?;
-    let channel_name = format!("{}_{}", pubkey_id, channel_type);
-    listener.listen(&channel_name).await?;
-    debug!("Listening on channel: {}", channel_name);
-
-    // Step 3: Wait for a notification and then fetch the new content
-    loop {
-        let notification = listener.recv().await?;
-        debug!("Received notification: {:?}", notification);
-        if notification.channel() == channel_name {
-            let row: (Vec<u8>,) = sqlx::query_as(&format!(
-                "SELECT {} FROM directory WHERE pubkey_id = $1",
-                channel_type
-            ))
-            .bind(pubkey_id)
-            .fetch_one(pool)
-            .await?;
-
-            let updated_content = row.0;
-            if !updated_content.is_empty() {
-                return Ok(updated_content);
+    pub async fn peek(&self, pubkey_id: &str, channel_type: &str) -> RedisResult<Vec<u8>> {
+        let mut conn = self.client.get_async_connection().await?;
+        let key = format!("{}:{}", pubkey_id, channel_type);
+
+        // Attempt to fetch existing content for the given pubkey_id and channel_type
+        if let Ok(data) = conn.get::<_, Vec<u8>>(&key).await {
+            if !data.is_empty() {
+                return Ok(data);
             }
         }
+        debug!("Failed to fetch content initially");
+
+        // Set up a temporary listener for changes
+        let mut pubsub_conn = self.client.get_async_connection().await?.into_pubsub();
+        let channel_name = format!("{}:{}", pubkey_id, channel_type);
+        pubsub_conn.subscribe(&channel_name).await?;
+
+        // Use a block to limit the scope of the mutable borrow
+        let data = {
+            let mut message_stream = pubsub_conn.on_message();
+
+            loop {
+                match message_stream.next().await {
+                    Some(msg) => {
+                        let _: () = msg.get_payload()?; // Notification received
+                                                        // Try fetching the data again
+                        if let Some(data) = conn.get::<_, Option<Vec<u8>>>(&key).await? {
+                            if !data.is_empty() {
+                                break data; // Exit the block, returning the data
+                            }
+                        }
+                    }
+                    None =>
+                        return Err(RedisError::from((
+                            ErrorKind::IoError,
+                            "PubSub connection closed",
+                        ))),
+                }
+            }
+        };
+
+        // Since the stream is dropped here, we can now unsubscribe
+        pubsub_conn.unsubscribe(&channel_name).await?;
+
+        Ok(data)
```

Comment:
I think its important testing the database implementation

Reply:
there aren't unit tests, but the integration tests do test all functions including async peek.